### ISSUE TITLE
Improved Provisioner Limit Metric/Alerting

### ISSUE THREAD

**Issue by sidewinder12s** [kubernetes-sigs/karpenter#676]
### Description

**What problem are you trying to solve?**

I need a clear way to alert when provisioners hit their limits and will no longer provision capacity. We've tried setting some metric alerts based on % of the limit provisioned but this has turned out to be tricky. 

1. Tried setting an alert for 100% utilization. This did not work as we'd almost never hit 100% because of the sizes of instances would never match up exactly with the limit
2. Tried setting an alert for 90% utilization. This has also failed in a couple cases where new workloads came in requiring significant capacity and when the provisioner was already at 80% it could not provision more capacity. 

The latter problem is even worse if you have a diverse set of workloads of mixed sizes. Some of our clusters we can slowly creep up to 90+ percent utilization with lots of smaller workloads so the alerts work as intended, other times we'll have large pods come in that can't provision capacity and will be blocked. 

I think it'd be easier/simpler to operate if there was just another metric per provisioner that just said I am currently limited, in addition to metric that currently exists.

Another option might be to just alert based on pods failing to schedule, but that will likely produce its own issues with the multitude of ways that pods can fail to schedule.

**How important is this feature to you?**

Fairly important operational consideration.

Related issue around how hard it is to tell if you've hit the provisioner limit: https://github.com/aws/karpenter-core/issues/686

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by gfcroft** [comment]
Note: provisioners are being changed to to "nodepools" - see: https://karpenter.sh/docs/upgrading/v1beta1-migration/#provisioner---nodepool

@sidewinder12s it seems that we could create a metric on limits being reached whenever this limit function returns an error: https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/apis/v1beta1/nodepool.go

Of course that would mean that you would have to have at least one occasion where a node has failed to be launched before you would actually have any metric produced which you could use to alert on - is this what you were thinking of? 

Additionally, would it be particularly helpful for there to be more specific metrics on the exact limit that was reached? e.g. counts of memory limit threshold exceeded for node group, alongside an all-encompassing count of "limit exceeded" metric for  a given node pool

**Comment by ellistarn** [comment]
I'd love to pop a status condition for this, and then generate metrics and events from the status condition.

**Comment by gfcroft** [comment]
@ellistarn sounds interesting -  presumably you could use this new framework to handle the related issue (https://github.com/kubernetes-sigs/karpenter/issues/705) too? Looking over it briefly it seems it could work well here

**Comment by ellistarn** [comment]
Yes. I'd like to see what about karpenter's operations can't be modeled in this way.  I'm hopeful that we can cover almost everything with this mechanism.

**Comment by den-is** [comment]
We have dozens of clusters and thousands of nodes.
Such an easy metric "on reaching limits", would make life so much easier. <3
